File: src/prefab/resolve.ts
import { Package, PackageRequirement, Installation } from "types"
import { useCellar, useInventory } from "hooks"
import { TeaError } from "utils"

/// NOTE resolves to bottles
/// NOTE contract there are no duplicate projects in input

interface RT {
  /// fully resolved list (includes both installed and pending)
  pkgs: Package[]

  /// already installed packages
  installed: Installation[]

  /// these are the pkgs that aren’t yet installed
  pending: Package[]
}

/// resolves a list of package specifications based on what is available in
/// bottle storage if `update` is false we will return already installed pkgs
/// that resolve so if we are resolving `node>=12`, node 13 is installed, but
/// node 19 is the latest we return node 13. if `update` is true we return node
/// 19 and *you will need to install it*.
export default async function resolve(reqs: (Package | PackageRequirement)[], {update}: {update: boolean} = {update: false}): Promise<RT> {
  const inventory = useInventory()
  const cellar = useCellar()
  const rv: RT = { pkgs: [], installed: [], pending: [] }
  let installation: Installation | undefined
  for (const req of reqs) {
    if (!update && (installation = await cellar.has(req))) {
      // if something is already installed that satisfies the constraint then use it
      rv.installed.push(installation)
      rv.pkgs.push(installation.pkg)
    } else {
      const version = await inventory.select(req)
      if (!version) {
        throw new TeaError("not-found: pkg.version", {pkg: req})
      }
      const pkg = { version, project: req.project }
      rv.pkgs.push(pkg)

      if ((installation = await cellar.has(pkg))) {
        rv.installed.push(installation)
      } else {
        rv.pending.push(pkg)
      }
    }
  }
  return rv
}

File: crates/lib/src/resolve.rs
use crate::config::Config;
use crate::types::{Installation, Package, PackageReq};
use crate::{cellar, inventory};
use std::error::Error;

#[derive(Debug, Default)]
pub struct Resolution {
    /// fully resolved list (includes both installed and pending)
    pub pkgs: Vec<Package>,

    /// already installed packages
    pub installed: Vec<Installation>,

    /// these are the pkgs that aren’t yet installed
    pub pending: Vec<Package>,
}

//TODO no need to take array since it doesn’t consider anything
use futures::stream::{FuturesUnordered, StreamExt};

pub async fn resolve(reqs: Vec<PackageReq>, config: &Config) -> Result<Resolution, Box<dyn Error>> {
    let mut rv = Resolution::default();

    // Create a FuturesUnordered to run the tasks concurrently
    let mut futures = FuturesUnordered::new();

    for req in reqs {
        futures.push(async move {
            if let Some(installation) = cellar::has(&req, config).await {
                Ok::<_, Box<dyn Error>>((
                    Some((installation.clone(), installation.pkg.clone())),
                    None,
                ))
            } else if let Ok(Some(version)) = inventory::select(&req, config).await {
                let pkg = Package {
                    project: req.project.clone(),
                    version,
                };
                Ok::<_, Box<dyn Error>>((None, Some(pkg)))
            } else {
                Err(Box::new(ResolveError { pkg: req }) as Box<dyn Error>)
            }
        });
    }

    // Process the results as they are completed
    while let Some(result) = futures.next().await {
        match result? {
            (Some((installation, pkg)), None) => {
                rv.installed.push(installation);
                rv.pkgs.push(pkg);
            }
            (None, Some(pkg)) => {
                rv.pkgs.push(pkg.clone());
                rv.pending.push(pkg);
            }
            _ => unreachable!(), // This should not happen
        }
    }

    Ok(rv)
}

use std::fmt;

#[derive(Debug)]
pub struct ResolveError {
    pub pkg: PackageReq, // Holds the package or requirement
}
impl Error for ResolveError {}

impl fmt::Display for ResolveError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "not-found: pkg: {:?}", self.pkg)
    }
}
